### ISSUE TITLE
Some feedback after trying it out

### ISSUE THREAD

**Issue by pjrulez** [rick-jennings/phable#11]
My use case requires me to periodically authenticate and then traverse a number of projects and read some records from historian.

- It would be great if there was a way to send in an array of p:proj123:r:record123 ids, and then the client would be calling/authenticating into each project based on p:proj123 instead of requiring separate calls for each project.

I'm trying to use this library like this:
```
end = date.today()
start = end - timedelta(days=1)
# range = DateTimeRange(start, end)
range = DateRange(start, end)
#pprint(str(range))
counter = 1
#with open_haxall_client(uri=settings.TIPIFY_URI.geturl(), username=settings.TIPIFY_USER, password=settings.TIPIFY_PASS) as client:
# with (open_haystack_client(uri=settings.TIPIFY_URI.geturl(), username=settings.TIPIFY_USER, password=settings.TIPIFY_PASS) as client):
#     pprint(f"skyspark_test :: initial authentication :: url: {settings.TIPIFY_URI.geturl()} :: About: {client.about()}")

for proj, rec in zip(pids_list[0:], rids_list[0:]):
    try:
        pprint(f"skyspark_test :: counter: {counter} :: proj: {proj} :: url: {settings.TIPIFY_BASE_URI.geturl()}/{proj}")
        with open_haystack_client(f"{settings.TIPIFY_BASE_URI.geturl()}/{proj}",
                             username=settings.TIPIFY_USER,
                             password=settings.TIPIFY_PASS) as client:
            #result = client.call(f"{proj}/hisRead", data=Grid.to_grid({"id": Ref(val=f"p:{proj}:{rec}"), "range": range})).to_pandas_all()
            #result = client.his_read_by_id(id=Ref('p:t245g_b21190228:r:2a9907f9-b6e2001e'), range=range)
            result = client.his_read_by_id(id=Ref(val=f"p:{proj}:{rec}"), range=range).to_pandas_all()
            #client.close()

            pprint(f"#{counter} - {settings.TIPIFY_BASE_URI.geturl()}/{proj} :: result: ")
            pprint(result[0])
            pprint(result[1])
    except Exception as e:
        pprint(f"skyspark_test :: counter: {counter} :: proj: {proj} :: rec: {rec} :: error: {e}")
        pprint(f'Traceback: {traceback.format_exc()}')
    finally:
        counter += 1
```

- I'm receiving a lot of "Expected status 200 and received status 400" errors, or connection interruptions.
_raise IncorrectHttpResponseStatus("phable.http.IncorrectHttpResponseStatus: ('Expected status 200 and received status 404.', 404)")_

Pretty much every other his_read_by_id call breaks like this.

- There are some issues with '_create_his_read_req_data' method. It's not handling DateRange | DateTimeRange kinds properly. The constructed request Grid should inlude "from" and "to" nested under "range". In the case of DateTimeRange, there needs to also be a timezone "tz" field.

- Also, I've been seeing some exceptions that arose from the server failing to parse ids when a list of Ref kind objects is passed to 'his_read_by_ids' method. It appears it also originates from  '_create_his_read_req_data'

- Often, my first call to 'open_haystack_client' will result in:
'error: [WinError 10054] An existing connection was forcibly closed by the remote host'

Thanks, that's all I've got for now...

[example.log](https://github.com/user-attachments/files/17597595/example.log)

**Comment by rick-jennings** [comment]
Thanks for the feedback.  Below is guidance on various topics you mentioned.

I will keep this issue open since I think we can improve Phable's logging and error messages in a future release to make debugging easier.

## Topic 1

> It would be great if there was a way to send in an array of p:proj123:r:record123 ids, and then the client would be calling/authenticating into each project based on p:proj123 instead of requiring separate calls for each project.

Typically I recommend using the context manager `open_haxall_client` to keep it simple.  For this use case you might consider managing a `HaxallClient` object with the uri associated each project database using the `open` and `close` methods.

If SkySpark is involved, I would suggest you consider leveraging the platform's clustering features using `open_haxall_client` and axon code evaluated using `HaxallClient.eval()`.

## Topic 2

> There are some issues with '_create_his_read_req_data' method. It's not handling DateRange | DateTimeRange kinds properly. The constructed request Grid should inlude "from" and "to" nested under "range". In the case of DateTimeRange, there needs to also be a timezone "tz" field.

The `datetime` objects used in `DateTimeRange` must use `ZoneInfo` to define the timezone.  I will update the docs to make this more clear.  Also, we could consider adding validation to `DateTimeRange` to raise an error if the embedded datetime objects do not define the timezone correctly.

Here is an example:

```python
from datetime import datetime
from zoneinfo import ZoneInfo

ts_now = datetime.now(ZoneInfo("America/New_York"))
```

## Topic 3

> Also, I've been seeing some exceptions that arose from the server failing to parse ids when a list of Ref kind objects is passed to 'his_read_by_ids' method. It appears it also originates from '_create_his_read_req_data'

Please let me know if you see a `CallError` or `UnknownRecError` raised by Phable.  If applicable, in the meantime you could catch the `CallError` object and see what's within the `help_msg` described [here](https://phable.dev/haystack_client/#phable.haystack_client.CallError).

Also, let's see if the feedback in Topic 1 mentioned above helps address this problem.

## Topic 4

> Often, my first call to 'open_haystack_client' will result in:
'error: [WinError 10054] An existing connection was forcibly closed by the remote host'

This might be a networking issue unrelated to Phable.  Please reach out to the organization that is supporting your SkySpark instance if applicable.

**Comment by pjrulez** [comment]
Thanks for your feedback. It's very helpful.
I switched to using .eval method with xq() and axon.
Also, in that case it's good to know that references need to be prefixed with @ and simply listed without any quotes. This is how I could access all of the necessary projects as once.
I didn't know that there's a whole lot of parsing going on under the hood. I switched to using ZoneInfo. Thanks.

.eval method return a phable Grid. I'm now checking if using Grid or .to_pandas_all() is easier or more efficient.

When I can play around with it a bit again, I'll see what's going on with UnknownRecError because I definitely saw some!

**Comment by rick-jennings** [comment]
Thanks for the feedback.

> I didn't know that there's a whole lot of parsing going on under the hood.

There is a decent amount of parsing.  ‘ZoneInfo’ usage helps us map to a Haystack defined time zone in a cleaner way.  As I mentioned I want to raise an error in the future if the ‘datetime’ object does not have the necessary ‘ZoneInfo’ which should improve the API user experience.

FYI, recently we simplified how we convert from a Haystack Grid to a Pandas or Polars dataframe considerably.  A lot of thought went into that and I would be interested in your feedback.

Take a look at the docs for ‘UnknownRecError’.  Probably that is not a related Phable issue.  You might want to look into using the ‘checked’ parameter for the ‘read’ or ‘read_by_id’ methods on ‘Client’.

Lastly, based on your initial post I plan to allow for custom headers in an upcoming release, which might help with Topic 4 mentioned in my last post.

**Comment by rick-jennings** [comment]
As a follow up to my earlier message -

The just released version 0.1.17 of phable may help address some of these issues.

## Revisiting Topic 2

Post init validation has been added to `DateTimeRange`.  Also, the docs have been improved and now include usage examples for `DateTimeRange`.

## Revisiting Topic 4

After analyzing several related issues that were recently reported, we found that some cloud services expect specific criteria for a `User-Agent` header in the HTTP request.  We updated or added the `User-Agent` header used in HTTP requests made using Phable and proved that resolved an issue with at least one cloud service.

## Other discussion

In my last message I referred to a `checked` parameter specific to Phable's API.  Please note the SkySpark API has a similar feature which may be used in your Axon code evaluated by Phable.  For more details on that please review the Axon documentation by SkyFoundry.

I am going to close this issue now.  If you continue experiencing an issue described in `Topic 4` after trying phable v0.1.17, please create another issue focused on that.  Also, you are welcome to open other issues too.

Thanks for reporting this issue and helping to improve Phable!
